Tests: Virtual Machines. Покрытие API модуля тестами

dr-wh0 opened the pull request:
## Описание изменений

Реализованы **интеграционные тесты** API модуля виртуальных машин (_**Virtual Machines**_) для всех 14 эндпоинтов. Полное покрытие включает создание, удаление, управление состоянием, клонирование, редактирование, работу со снапшотами и доступ к VNC. Тесты используют как существующие фикстуры проекта, так добавляют новые специализированные фикстуры для ВМ. Покрыты все основные сценарии работы, включая валидацию данных, обработку ошибок и edge cases.

## Ключевые изменения

**Тестовая инфраструктура:**
- Фикстуры в `conftest.py` проекта:
  - `storage`, `volume`, - для создания, клонирования и редактирования ВМ.
  - `vm_create_data` - возвращает базовые данные для создания ВМ
  - `vm_edit_data` - возвращает данные для редактирования ВМ  
  - `virtual_machine` - создает тестовую ВМ и удаляет после теста
  - `deactivated_virtual_machine` - создает выключенную ВМ
  - `activated_virtual_machine` - создает запущенную ВМ
- Добавлены фикстуры в `conftest.py` модуля виртуальных машин:
  - `cleanup_vms` - очищает все ВМ и связанные ресурсы до и после каждого теста
  - `vm_snapshot` - создает снапшот ВМ и удаляет после теста

**Утилиты (новые):**
- `cleanup_all_virtual_machines()` - удаляет все ВМ из БД и libvirt
- `check_object_exists()` - проверяет существование объекта по ID в API

## Реализованные тесты

**Общая структура файлов для тестов:**
```
openvair/modules/virtual_machines/tests/
├── conftest.py
└── api/
    ├── test_create.py
    ├── test_delete.py
    ├── test_edit.py
    ├── test_manage.py
    ├── test_retrieve.py
    └── snapshots/
        ├── test_snap_create.py
        ├── test_snap_delete.py
        ├── test_snap_manage.py
        └── test_snap_retrieve.py
```

### 1. Создание (и клонирование) ВМ (`test_create.py`)
- `test_create_vm_success` - успешное создание ВМ
- `test_create_vm_auto_volume_success` - создание с автоматическим созданием тома
- `test_create_vm_missing_fields` - создание без обязательных полей
- `test_create_vm_duplicate_name` - создание с дублирующимся именем
- `test_create_vm_nonexistent_volume` - создание с несуществующим томом
- `test_create_vm_invalid_graphic_interface` - невалидный графический интерфейс
- `test_create_vm_invalid_cpu` - невалидная конфигурация CPU
- `test_create_vm_invalid_ram` - невалидный размер RAM
- `test_create_vm_invalid_os` - невалидная конфигурация ОС
- `test_create_vm_invalid_mac` - невалидный MAC-адрес
- `test_create_vm_invalid_virtual_interface` - невалидный сетевой интерфейс
- `test_create_vm_invalid_disks` - невалидная конфигурация дисков
- `test_create_vm_invalid_name` - невалидное имя ВМ
- `test_create_vm_empty_disks` - создание без дисков
- `test_create_vm_unauthorized` - неавторизованный запрос создания ВМ
- `test_clone_vm_success` - успешное клонирование ВМ
- `test_clone_vm_multiple_success` - успешное множественное клонирование
- `test_clone_vm_invalid_count` - невалидное количество клонов
- `test_clone_vm_nonexistent_vm` - клонирование несуществующей ВМ
- `test_clone_vm_nonexistent_storage` - клонирование на несуществующее хранилище
- `test_clone_vm_running_vm` - клонирование запущенной ВМ
- `test_clone_vm_missing_target_storage` - клонирование без целевого хранилища
- `test_clone_vm_invalid_uuid` - невалидный UUID ВМ
- `test_clone_vm_invalid_target_storage_uuid` - невалидный UUID хранилища
- `test_clone_vm_unauthorized` - неавторизованный запрос клонирования

### 2. Удаление ВМ (`test_delete.py`)
- `test_delete_vm_success` - успешное удаление ВМ
- `test_delete_vm_invalid_uuid` - невалидный UUID
- `test_delete_vm_running` - удаление запущенной ВМ
- `test_delete_vm_nonexistent` - удаление несуществующей ВМ
- `test_delete_vm_invalid_statuses` - удаление ВМ в невалидных статусах
- `test_delete_vm_with_snapshots_success` - удаление ВМ со снапшотами
- `test_delete_vm_unauthorized` - неавторизованный запрос удаления

### 3. Редактирование ВМ (`test_edit.py`)
- `test_edit_vm_success` - успешное редактирование ВМ
- `test_edit_vm_add_disk_success` - успешное редактирование с добавлением диска (тома)
- `test_edit_vm_invalid_uuid` - невалидный UUID
- `test_edit_vm_nonexistent` - редактирование несуществующей ВМ
- `test_edit_vm_invalid_payload` - невалидные данные редактирования
- `test_edit_vm_running` - редактирование запущенной ВМ
- `test_edit_vm_duplicate_name` - изменение на дублирующееся имя
- `test_edit_vm_unauthorized` - неавторизованный запрос редактирования

### 4. Управление состоянием ВМ - запуск, выключение, доступ к VNC (`test_manage.py`)
- `test_start_vm_success` - успешный запуск ВМ
- `test_start_vm_already_running` - запуск уже запущенной ВМ
- `test_start_vm_invalid_uuid` - невалидный UUID при запуске
- `test_start_vm_nonexistent` - запуск несуществующей ВМ
- `test_start_vm_unauthorized` - неавторизованный запуск
- `test_shut_off_vm_success` - успешная остановка ВМ
- `test_shut_off_vm_already_shutoff` - остановка уже остановленной ВМ
- `test_shut_off_vm_invalid_uuid` - невалидный UUID при остановке
- `test_shut_off_vm_nonexistent` - остановка несуществующей ВМ
- `test_shut_off_vm_unauthorized` - неавторизованный запрос остановки
- `test_vnc_vm_success` - успешный доступ к VNC
- `test_vnc_vm_shutoff` - VNC доступ к остановленной ВМ
- `test_vnc_vm_nonexistent` - VNC доступ к несуществующей ВМ
- `test_vnc_vm_invalid_uuid` - невалидный UUID для VNC
- `test_vnc_vm_unauthorized` - неавторизованный доступ
- `test_vm_restart_with_snapshot` - перезапуск ВМ со снапшотом

### 5. Получение ВМ (`test_retrieve.py`)
- `test_get_all_vms_success` - получение списка ВМ
- `test_get_vm_by_id_success` - получение ВМ по ID
- `test_get_vm_validate_structure_success` - проверка структуры данных ВМ
- `test_get_all_vms_empty_list_success` - получение пустого списка ВМ
- `test_get_all_vms_with_pagination` - проверка пагинации списка ВМ
- `test_get_nonexistent_vm_by_id` - получение несуществующей ВМ
- `test_get_vm_invalid_uuid` - невалидный UUID при получении
- `test_get_all_vms_unauthorized` - неавторизованный доступ к списку
- `test_get_vm_by_id_unauthorized` - неавторизованный доступ к конкретной ВМ

### 6. Снапшоты: Создание (`test_snap_create.py`)
- `test_create_snapshot_success` - успешное создание снапшота
- `test_create_snapshot_missing_name` - создание без имени
- `test_create_snapshot_duplicate_name` - создание с дублирующимся именем
- `test_create_snapshot_shutoff_vm` - создание на остановленной ВМ
- `test_create_snapshot_nonexistent_vm` - создание на несуществующей ВМ
- `test_create_snapshot_invalid_name` - невалидное имя снапшота
- `test_create_snapshot_no_description` - создание без описания
- `test_create_snapshot_invalid_vm_uuid` - невалидный UUID ВМ
- `test_create_multiple_snapshots_chain` - создание цепочки снапшотов
- `test_create_snapshot_unauthorized` - неавторизованный запрос создания снапшота

### 7. Снапшоты: Удаление (`test_snap_delete.py`)
- `test_delete_snapshot_success` - успешное удаление снапшота
- `test_delete_snapshot_shutoff_vm_success` - удаление из остановленной ВМ
- `test_delete_snapshot_chain_success` - проверка восстановления цепочки при удалении снапшота
- `test_delete_snapshot_invalid_vm_uuid` - невалидный UUID ВМ
- `test_delete_snapshot_invalid_snapshot_uuid` - невалидный UUID снапшота
- `test_delete_snapshot_nonexistent_vm` - удаление из несуществующей ВМ
- `test_delete_snapshot_nonexistent_snapshot` - удаление несуществующего снапшота
- `test_delete_snapshot_wrong_vm` - удаление с неправильным ID ВМ
- `test_delete_snapshot_unauthorized` - неавторизованный запрос удаления снапшота

### 8. Снапшоты: Управление (`test_snap_manage.py`)
- `test_revert_snapshot_success` - успешный откат к снапшоту
- `test_revert_snapshot_chain_success` - откат в цепочке снапшотов
- `test_revert_snapshot_shutoff_vm` - откат для остановленной ВМ
- `test_revert_snapshot_nonexistent_vm` - откат для несуществующей ВМ
- `test_revert_snapshot_nonexistent_snapshot` - откат к несуществующему снапшоту
- `test_revert_snapshot_invalid_vm_uuid` - невалидный UUID ВМ
- `test_revert_snapshot_invalid_snapshot_uuid` - невалидный UUID снапшота
- `test_revert_snapshot_wrong_vm` - откат с неправильным ID ВМ
- `test_revert_snapshot_unauthorized` - неавторизованный запрос отката снапшота

### 9. Снапшоты: Получение (`test_snap_retrieve.py`)
- `test_get_all_snapshots_success` - получение всех снапшотов ВМ
- `test_get_snapshot_by_id_success` - получение снапшота по ID
- `test_get_all_snapshots_empty_list_success` - получение пустого списка снапшотов
- `test_get_snapshot_nonexistent_vm` - получение для несуществующей ВМ
- `test_get_snapshot_nonexistent_snapshot` - получение несуществующего снапшота
- `test_get_snapshot_invalid_vm_uuid` - невалидный UUID ВМ
- `test_get_snapshot_invalid_snapshot_uuid` - невалидный UUID снапшота
- `test_get_snapshot_wrong_vm` - получение с неправильным ID ВМ
- `test_get_all_snapshots_unauthorized` - неавторизованный доступ к списку снапшотов
- `test_get_snapshot_by_id_unauthorized` - неавторизованный доступ к конкретному снапшоту


## Особенности

- Тесты используют существующие фикстуры проекта (`storage`, `volume` и др.), что обеспечивает согласованность тестовой среды
- В конфигурации `.env.tests` требуется лишь настройка переменных `TEST_STORAGE_PATH` и `TEST_STORAGE_FS_TYPE` для создания `storage` и `volume` для виртуальных машин.
- Для проверки состояний ВМ используются утилиты ожидания, учитывающие асинхронную природу операций с виртуальными машинами
- Интеграция с libvirt для проверки реального состояния ВМ и снапшотов на уровне гипервизора
- Очистка тестовых данных включает удаление ВМ из БД и libvirt, а также связанных ресурсов
- Тесты покрывают сложные сценарии: цепочки снапшотов, клонирование, редактирование конфигурации
- Все эндпоинты проверены на неавторизованный доступ
- Добавлена [документация](https://github.com/Aerodisk/openvair-docs/pull/19) к тестам

## Критерии приёмки

- [x] Создать структуру папок для тестов
- [x] Реализовать все тестовые сценарии для основных операций
- [x] Добавить фикстуры для создания тестовых ВМ и зависимостей
- [x] Покрыть основные ошибки и edge cases
- [x] Интегрировать с существующей тестовой инфраструктурой
- [ ] Обеспечить стабильное прохождение всех тестов **(WIP)**
- [x] Добавить документацию к тестам ([документация](https://github.com/Aerodisk/openvair-docs/pull/19))

## Связанные задачи

- Issue #272

dr-wh0 committed:
Add create VM integration API tests

dr-wh0 committed:
Add delete VM integration API tests

dr-wh0 committed:
Add retrieve VM integration API tests

dr-wh0 committed:
Add manage VM integration API tests

dr-wh0 committed:
Add edit VM integration API tests

dr-wh0 committed:
Merge dev into tests/virtual_machines

dr-wh0 committed:
Merge branch 'dev' into tests/virtual_machines

dr-wh0 committed:
Add storage cleanup in cleanup_vms() fixture

dr-wh0 committed:
Add create clone VM integration API tests

dr-wh0 committed:
Add VM snapshots integration API tests

dr-wh0 committed:
Fix DB concurrent update error in base_manager

dr-wh0 committed:
Add cascade delete snapshots in database in ORM

dr-wh0 committed:
Revert "Fix DB concurrent update error in base_manager"

This reverts commit 3180b9e84dcef0316e8c1831082d48a88ba2c94b.

dr-wh0 committed:
Fix VM API (str -> UUID) and schemas validation

dr-wh0 committed:
Refactor VM tests structure

dr-wh0 committed:
Refactor: ruff format VM tests

VladimirTrue commented on openvair/libs/testing/utils.py:
```diff
@@ -582,3 +608,34 @@ def cleanup_test_bridges() -> None:  # noqa: C901 because it will be simplified
         LOG.warning(f'Error during test bridges cleanup: {e}')
+
+
+def check_object_exists(
```
А на сколько нужна данная фукнкция? Вижу, что она, используется в фикстурах. А что мешает просто брать и удалять, т.к. мы подразумеваем то, что мы создали объект ранее

VladimirTrue commented on openvair/modules/virtual_machines/entrypoints/schemas.py:
```diff
@@ -111,7 +111,7 @@ class QOS(BaseModel):
     """Schema for disk information."""
 
-    name: Optional[str] = None
+    name: Optional[str] = ""
```
Почему замена на такое дефолтное значение? И если оно приемлемо, то почему сохраняем Optional?

VladimirTrue commented on openvair/modules/virtual_machines/tests/api/virtual_machines/test_edit.py:
```diff
@@ -0,0 +1,182 @@
+        'status',
+        'available',
+    )
+    wait_for_field_value(
```
А почему есть необходимость после ожидания availabale для status проверять тут `name` и далее `description`?

tihon49 replied:
@VladimirTrue resolved?

VladimirTrue reviewed (commented):
Всё гуд. Тесты пока не запускал. Ожидаю разрешения по ревью, после чего буду проверять работоспособность и апрувить

tihon49 commented on openvair/libs/testing/utils.py:
```diff
@@ -582,3 +608,34 @@ def cleanup_test_bridges() -> None:  # noqa: C901 because it will be simplified
+    if isinstance(response, Dict):
+        objects = response.get("items", response)
+        if isinstance(objects, Dict):
+            objects = [objects]
```
выглядит странно )))
может можно упростить, к примеру, как-то так:
```python
objects = list(response.get("items", response))
```
просто как вариант.

tihon49 commented on openvair/modules/virtual_machines/entrypoints/api.py:
```diff
@@ -369,7 +369,7 @@ async def get_snapshot(
     )
     snapshot = await run_in_threadpool(
-        crud.get_snapshot, str(vm_id), str(snap_id), user_info
+        crud.get_snapshot, vm_id, str(snap_id), user_info
```
`vm_id` - UUID, а `snap_id` - приводим к строке, хотя в `352 строке` видим что это тоже UUID. Почему так?

tihon49 commented on openvair/modules/virtual_machines/tests/conftest.py:
```diff
@@ -0,0 +1,84 @@
+    cleanup_all_images()
+    cleanup_all_volumes()
+    cleanup_all_templates()
+    cleanup_all_storages()
```
Зачем все очищать "до" теста?

Подозревается что тесты запускаются в чистой системе. Получаем избыточность и замедление тестов. Возможно я ошибаюсь.

dr-wh0 committed:
Refactor: Fix review
- Fix UUID usage in get_snapshot API
- Delete cleanups before each test
- Refactor cleanup_all_virtual_machines

dr-wh0 committed:
Fix str(snap_id) in get_snapshot

dr-wh0 committed:
Refactor: Fix review
- Add more edit_data asserts in test_edit_vm_success

dr-wh0 committed:
Refactor: Fix review
- Delete check_object_exists() util

dr-wh0 committed:
Test commit (minor docstrings refactor)

dr-wh0 committed:
Add SNAPSHOT_TIMEOUT constraint

dr-wh0 committed:
Fix vm creation with duplicate name

dr-wh0 committed:
Add mac validation in VM API schema

dr-wh0 committed:
Change available_power_states in edit_vm to only "shut_off"

dr-wh0 committed:
Add checking status and power_state before starting VM in start_vm()

dr-wh0 committed:
Fix asserts in test_edit_vm_success
